### ISSUE TITLE
`weight_decay` etc. works contradictory to `params` without grad

### ISSUE THREAD

**Issue by ILCSFNO** [pytorch/pytorch#150610]
### 🐛 Describe the bug

The docs of [torch.optim.Adam()](https://pytorch.org/docs/stable/generated/torch.optim.Adam.html#torch.optim.Adam), [torch.optim.AdamW()](https://pytorch.org/docs/stable/generated/torch.optim.AdamW.html#torch.optim.AdamW) and [torch.optim.RAdam()](https://pytorch.org/docs/stable/generated/torch.optim.RAdam.html#torch.optim.RAdam) show their shared description as below:

https://github.com/pytorch/pytorch/blob/9e106019f64d668f17f0b50dc46192cff7a37dce/torch/optim/adam.py#L41

and `weight_decay` is described as:

https://github.com/pytorch/pytorch/blob/9e106019f64d668f17f0b50dc46192cff7a37dce/torch/optim/adam.py#L322

For repro below:
### Repro
```python
import torch
import numpy as np
input_data = np.random.rand(10, 10)
input_tensor = torch.from_numpy(input_data)
input_tensor = torch.autograd.Variable(input_tensor, requires_grad=False)
optimizer = torch.optim.Adam([{'params': input_tensor, 'weight_decay': 0.01}])
# optimizer = torch.optim.AdamW([{'params': input_tensor, 'weight_decay': 0.01}])
# optimizer = torch.optim.RAdam([{'params': input_tensor, 'weight_decay': 0.01}])
optimizer.step()
```

It can run well.

But when we focus on the two parameters: 
* `requires_grad` in `torch.autograd.Variable`
* `weight_decay` in `torch.optim.Adam`, `torch.optim.AdamW` and `torch.optim.RAdam`

We notice that `params` with `requires_grad=False` is compatible with `weight_decay` set non-zero. However, it shouldn't!

As we all know, `weight_decay` is only useful when grad is allowed. The optimizer attempts to apply `weight_decay` to `input_tensor` which has `requires_grad=False`. Adding `weight_decay` to a parameter with `requires_grad=False` creates invalid gradient modification, which should lead to an error when trying to modify non-trainable parameters during the optimization step.

### Suggestions For Discussion above (Seemed Painful)
* Set initial value of weight_decay to None instead of 0
* Raise ValueError or UserWarning when `weight_decay` is set and `params` not require grad
e.g.
```txt
`params` should require grad when `weight_decay` is set
```

### Further
While we focus on the `weight_decay`, more other arguments should be noted too, however, it's painful to fix yet.

Look back to `Optimizer`, it is a crucial component in deep learning training, responsible for updating model parameters based on the gradient of the loss function, thereby continuously improving model performance.

So another useful path is to reject `params` without grad needed, that is, suggestions below:
### Suggestions Modified (Not Painful)
* Check `requires_grad` parameter of argument `params`, if True, accepted, otherwise, rejected and raise ValueError.
* Add Warning in description of `Optimizer`, from:

https://github.com/pytorch/pytorch/blob/9e106019f64d668f17f0b50dc46192cff7a37dce/torch/optim/optimizer.py#L335-L348

to:

```python
class Optimizer:
    r"""Base class for all optimizers.

    .. warning::
        Parameters need to be specified as collections that have a deterministic
        ordering that is consistent between runs. Examples of objects that don't
        satisfy those properties are sets and iterators over values of dictionaries.

    Args:
        params (iterable): an iterable of :class:`torch.Tensor` s or
            :class:`dict` s. Specifies what Tensors should be optimized,
            which should have grad available.
        defaults: (dict): a dict containing default values of optimization
            options (used when a parameter group doesn't specify them).
    """
```

Thanks for noting.

### Versions

Nightly

cc @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Varal7 @xmfan @vincentqb @jbschlosser @janeyx99 @crcrpar

**Comment by janeyx99** [comment]
Our optimizers only apply the update to parameters with grads (p.grad is not None): https://github.com/pytorch/pytorch/blob/main/torch/optim/adam.py#L150

So this problem should never occur unless someone has manually set requires_grad=False but allowed the parameter to have a grad attribute anyway. Are you running into such a case?

**Comment by ILCSFNO** [comment]
Oh thanks, it's awkward that I haven't note the filter of the grad in codes. I met with such a case just in a random try. Thanks a lot!
